osu.Game/Overlays/KeyBinding/KeyBindingRow.cs

```diff
@@ -150,6 +152,9 @@ protected override void OnHoverLost(HoverLostEvent e)
 
         private KeyButton bindTarget;
 
+        private ICollection<InputKey> previouslyHeldKeys;
+        private bool canHandleRelease;
```

Comment:
If I'm not mistaken, this flag is not necessary provided we save the state of buttons at the moment the binding component becomes focused.

* If a button went up and it was in `previouslyHeldKeys` before, then exclude it from `previouslyHeldKeys` and do not finalize binding input.
* If a button went up and it wasn't in `previouslyHeldKeys` before, then finalise.

I might be wrong though because this logic is quite hard to keep in my head.

Reply:
the current implementation takes the KeyCombination.FromInputState() at the moment a non-modified button is pressed and saves that as the binding. the issue with that is that it'll record things such as axes being off-center and switches (like actual 2-position switches), some of which don't even have an off-position say, down position activates button 30 while up position activates button 31. high-end devices like to do this and even though I'm not planning to use these to play Osu, this would make me have to unplug my devices to make it work

Reply:
I understand all that, I thought I was proposing a simpler implementation of the same behaviour, and I'm still not entirely sure the two differ (as in I don't see a concrete case where the two would differ). I'll have a think about it with a fresher mind tomorrow - but honestly I'd like to see tests for the scenarios you describe regardless.

Reply:
I started out with the same thing you've said, however I had to work around the edge case of having a button that's being held before the container is focused, releasing and pushing it again would register an empty KeyCombination.

I'll  get some tests going.